File: bench/py/28_str_concat.py
# string building by repeated `+=` (CPython's in-place refcount==1 optimization)
#
# ⛔ THE LOOP LIVES IN A FUNCTION ON PURPOSE. The in-place append fast
# path fires only when the accumulator is a LOCAL (the interpreter needs
# the following STORE_FAST to prove refcount==1); at module level `s` is
# a dict-backed global and every += copies the whole string - O(n^2),
# measured 2 to 11+ MINUTES at scale 25 on WSL2 (page-fault churn) vs
# 0.11s in this form. The module-level spelling silently defeated the
# very optimization this bench exists to measure.
import sys
scale = int(sys.argv[1]) if len(sys.argv) > 1 else 1

def main():
    N = 50000 * scale
    s = ""
    for i in range(N):
        s += str(i)
        s += ","
    print("result:", len(s))

main()

File: bench/cpp/28_str_concat.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/* Faithful C++ of bench/my/28_str_concat.my: string building by repeated `+=`
 * (std::string, appended in place like MyLang's unaliased-string fast path). */
#include "bench.h"
#include <string>

int main(int argc, char **argv)
{
    long scale = bench_scale(argc, argv);

    long N = 50000L * scale;
    std::string s;
    for (long i = 0; i < N; i++) {
        s += std::to_string(i);
        s += ",";
    }

    printf("result: %ld\n", (long)s.size());
    return 0;
}
